cmd/github-ldap-user-group-creator/main.go

```diff
@@ -0,0 +1,215 @@
+		logrus.WithError(err).Fatal("could not parse args")
+	}
+	return opts
+}
+
+func (o *options) validate() error {
+	level, err := logrus.ParseLevel(o.logLevel)
+	if err != nil {
+		return fmt.Errorf("invalid log level specified: %w", err)
+	}
+	logrus.SetLevel(level)
+
+	if o.mappingFile == "" {
+		return fmt.Errorf("--mapping-file must not be empty")
+	}
+	return nil
+}
+
+const (
+	appCIContextName = string(api.ClusterAPPCI)
+	toolName         = "github-ldap-user-group-creator"
+	groupSuffix      = "-group"
+)
+
+func addSchemes() error {
+	if err := userv1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add userv1 to scheme: %w", err)
+	}
+	return nil
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	opts := parseOptions()
+
+	if err := opts.validate(); err != nil {
+		logrus.WithError(err).Fatal("failed to validate the option")
+	}
+
+	if err := addSchemes(); err != nil {
+		logrus.WithError(err).Fatal("failed to add schemes")
+	}
+
+	kubeconfigs, err := opts.kubernetesOptions.LoadClusterConfigs()
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to load kubeconfigs")
+	}
+
+	inClusterConfig, hasInClusterConfig := kubeconfigs[kube.InClusterContext]
+	delete(kubeconfigs, kube.InClusterContext)
+	delete(kubeconfigs, kube.DefaultClusterAlias)
+
+	if _, hasAppCi := kubeconfigs[appCIContextName]; !hasAppCi {
+		if !hasInClusterConfig {
+			logrus.WithError(err).Fatalf("had no context for '%s' and loading InClusterConfig failed", appCIContextName)
+		}
+		logrus.Infof("use InClusterConfig for %s", appCIContextName)
+		kubeconfigs[appCIContextName] = inClusterConfig
+	}
+
+	kubeConfig := kubeconfigs[appCIContextName]
+	appCIClient, err := ctrlruntimeclient.New(&kubeConfig, ctrlruntimeclient.Options{})
+	if err != nil {
+		logrus.WithError(err).Fatalf("could not create client")
+	}
+
+	clients := map[string]ctrlruntimeclient.Client{}
+	for cluster, config := range kubeconfigs {
+		cluster, config := cluster, config
+		if cluster == appCIContextName {
+			continue
+		}
+		client, err := ctrlruntimeclient.New(&config, ctrlruntimeclient.Options{})
+		if err != nil {
+			logrus.WithError(err).WithField("cluster", cluster).Fatal("could not create client for cluster")
+		}
+		clients[cluster] = client
+	}
+
+	clients[appCIContextName] = appCIClient
+
+	ctx := interrupts.Context()
+
+	mapping := func(path string) map[string]string {
+		logrus.WithField("path", path).Debug("Loading the mapping file ...")
+		bytes, err := ioutil.ReadFile(path)
+		if err != nil {
+			logrus.WithField("path", path).WithError(err).Debug("Failed to read file")
+			return nil
+		}
+		var mapping map[string]string
+		if err := yaml.Unmarshal(bytes, &mapping); err != nil {
+			logrus.WithField("string(bytes)", string(bytes)).WithError(err).Debug("Failed to unmarshal")
+			return nil
+		}
+		return mapping
+	}(opts.mappingFile)
+
+	if err := ensureGroups(ctx, clients, mapping, opts.dryRun); err != nil {
```

Comment:
since the tool will run once, it doesn't have to hold the mapping in an closure, since its going to marshal it once, right?

Reply:
No, I mean that you can just pass the `map[string]string` directly.

Reply:
```
func(path string, mapping *map[string]string)  {}
```
Do you want me to do something like ^?

Reply:
I mean that the `ensureGroups()` should get a `map[string]string` instead of a closure.

Reply:
You can just do 

```go

		logrus.WithField("path", opts.mappingFile).Debug("Loading the mapping file ...")
		bytes, err := ioutil.ReadFile(opts.mappingFile)
		if err != nil {
			logrus.WithField("path", opts.mappingFile).WithError(err).Debug("Failed to read file")
			return nil
		}
		var mapping map[string]string
		if err := yaml.Unmarshal(bytes, &mapping); err != nil {
			logrus.WithField("string(bytes)", string(bytes)).WithError(err).Debug("Failed to unmarshal")
			return nil
		}

	if err := ensureGroups(ctx, clients, mapping, opts.dryRun); err != nil {
```

Reply:
What about the `return`s?

Reply:
As we discussed in slack, I would just leave the code as it is.
I think it is just a different taste of code style, not correctness or performance.